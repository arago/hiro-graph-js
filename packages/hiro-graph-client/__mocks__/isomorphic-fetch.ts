/* eslint-env jest */
// mock of isomorphic-fetch for tests.
// fetch is a single function so we mock it with a
// mock fn.

// we need to require the real thing, as it mutates `global`, giving us access
// to Response
require.requireActual('isomorphic-fetch');

// @ts-ignore
const Response = global.Response;
const mockFn = jest.fn();

export { mockFn };

export default function fetch(...args: any[]) {
  // wrap the outcome of the mock function in a "new Response";
  const result = mockFn(...args);

  if (!result) {
    return Promise.resolve(new Response(JSON.stringify({ args })));
  }

  const [status, body] = result;

  if (body instanceof Error) {
    return Promise.reject(body);
  } else if (typeof body === 'string') {
    return Promise.resolve(new Response(body, { status }));
  } else {
    return Promise.resolve(new Response(JSON.stringify(body), { status }));
  }
}
