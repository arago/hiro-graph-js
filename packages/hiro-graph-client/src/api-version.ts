const APP_API_BASE = '/api/app/7.0';
const AUTH_API_BASE = '/api/auth/6.1';
const EVENTS_WS_API_BASE = '/api/events-ws/6.1';
const GRAPH_API_BASE = '/api/graph/7.1';
const GRAPH_WS_API_BASE = '/api/graph-ws/6.1';
const IAM_API_BASE = '/api/iam/6.1';
const KI_API_BASE = '/api/ki/6';
const VARIABLES_API_BASE = '/api/variables/6';

export {
  APP_API_BASE,
  AUTH_API_BASE,
  EVENTS_WS_API_BASE,
  GRAPH_API_BASE,
  GRAPH_WS_API_BASE,
  IAM_API_BASE,
  KI_API_BASE,
  VARIABLES_API_BASE,
};
