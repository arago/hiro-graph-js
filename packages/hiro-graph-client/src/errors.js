/**
 *  helper for errors.
 */
//we keep a list here.
const errors = {};

const is = (code) => (err) => err.code && err.code === code;
const factory = (code, defaultMsg) => {
    //side effect of creating this, populates an index.
    errors[code] = (msg = defaultMsg, error = new Error()) => {
        error.message = msg;
        error.code = code;

        return error;
    };

    return errors[code];
};

//this creates new errors for a given error code an caches them
export const create = (code, msg, error = new Error()) => {
    if (code in errors) {
        return errors[code](msg, error);
    }

    return factory(code, `Unknown Error (${code})`)(msg, error);
};

//export all the common errors.
export const notFound = factory(404, 'Not Found');
export const isNotFound = is(404);
export const forbidden = factory(403, 'Forbidden');
export const isForbidden = is(403);
export const unauthorized = factory(401, 'Unauthorized');
export const isUnauthorized = is(401);
export const badRequest = factory(400, 'Bad Request');
export const isBadRequest = is(400);
export const conflict = factory(409, 'Conflict');
export const isConflict = is(409);
export const transactionFail = factory(888, 'Transaction Failed');
export const isTransactionFail = is(888);
export const unknown = factory(500, 'Unknown Error');
//this is slightly different, no code, or code greater than or equal 500.
export const isUnknown = (err) => !err.code || err.code / 100 >= 5;

//another error is that the connection was closed before we got the chance to send data.
//this is a special case
export const connectionClosedBeforeSend = new Error(
    'Connection closed while waiting for token',
);
