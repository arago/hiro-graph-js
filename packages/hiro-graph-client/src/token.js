/**
 *  Represents a GraphIT Access Token.
 *
 *  Access is always a promise, as we may have to renew it.
 *
 *  both `onInvalidate()` and `getToken()` can return promises.
 *  `getMeta` should return synchronously.
 */

// note: this error is for "CANNOT" get a token, not "FAILED" to get a token.
// i.e. we look up in a database and it's not there. no amount of retries will
// succeed.
export const cannotGetToken = (reason) => {
    return Object.assign(new Error(reason), { cannotGetToken: true });
};

export default class Token {
    constructor({ onInvalidate = () => {}, getMeta = () => {}, getToken }) {
        this._onInvalidate = onInvalidate;
        this._getMeta = getMeta;
        this._getToken = getToken;
    }

    get() {
        if (!this._tokenPromise) {
            this._tokenPromise = Promise.resolve(this._getToken())
                .catch((err) => {
                    // flatten this, so another attempt can be made
                    this._tokenPromise = null;

                    // check if this is a permanent error
                    if (err.cannotGetToken) {
                        // we should throw. otherwise the user will end up in an infinite loop
                        throw err;
                    }

                    // otherwise return a bogus token
                    // returning a bogus token ("", falsy), should hopefully trigger this again,
                    // waiting subscribers will receive it, fail, and try again.
                    console.error('ERROR GETTING TOKEN!', err);

                    return '';
                })
                .then((token) => {
                    this._invalidated = false; // this resolved one way or another...

                    return token;
                });
        }

        return this._tokenPromise;
    }

    invalidate() {
        if (this._invalidated) {
            return;
        }

        this._tokenPromise = false;
        this._invalidated = true;

        return Promise.resolve(this._onInvalidate()).catch(() => {});
    }

    //returns nothing by default, but you may want to synchronously return the
    //last valid token, the time is was acquired, it's expiry, a refresh token, etc...
    meta() {
        return this._getMeta();
    }
}

export const fixedToken = (tokenString) =>
    new Token({ getToken: () => tokenString });
