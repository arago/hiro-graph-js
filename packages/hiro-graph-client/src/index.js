import Token, { cannotGetToken } from './token';
import Client from './client';
import * as Errors from './errors';

export default Client;
export { Token, cannotGetToken, Errors };
