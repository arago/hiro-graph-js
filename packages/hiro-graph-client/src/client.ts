/**
 *  A GraphIT Client
 *
 *  It will use Websockets where possible and fall back to HTTP where WebSockets not supported.
 *  Event streams require WebSockets so those will not work witout websocket support.
 *  The connection requires a `Token`.
 */
import { catchError, map, retryWhen, concatMap, delay } from 'rxjs/operators';
import { of, throwError, EMPTY } from 'rxjs';

import {
  WebSocketTransport,
  webSocketsAvailable,
  WebSocketRequestOptions,
} from './transport-websocket';
import { HttpTransport } from './transport-http';
import * as Errors from './errors';
import { fixedToken } from './token';
import { EventStream, EventStreamRequest } from './eventstream';
import {
  GraphTransport,
  GraphRequest,
  TimeseriesResponse,
  HistoryQueryOptions,
} from './types';
import {
  GremlinQueryFunction,
  GremlinQuery,
  gremlin as CreateGremlin,
} from './gremlin';
import { Lucene, LuceneQueryOptions, lucene as CreateLucene } from './lucene';

import { Token } from '.';

export interface ClientOptions {
  endpoint: string;
  token: string | Token;
}

export type TransportOrOptions = WebSocketRequestOptions | GraphTransport;

const isTransport = (t: TransportOrOptions): t is GraphTransport =>
  'request' in t && typeof t.request === 'function';

export class Client {
  private endpoint: string;
  private http: HttpTransport;
  private transport: GraphTransport;
  private token: Token;

  constructor(
    { endpoint, token }: ClientOptions,
    transportOrOptions: TransportOrOptions = {},
  ) {
    this.endpoint = endpoint;

    this.token = this.setToken(token);

    //create the transports.
    //the http is for http only endpoint, e.g. the /_ki/* and /_variable/* servlets
    this.http = new HttpTransport(endpoint);

    // This is the main transport for most of the api.
    // Use Websocket if available (except if specifically disabled)
    // fallback to http, or use given transport
    // duck type the final parameter
    if (isTransport(transportOrOptions)) {
      this.transport = transportOrOptions;
    } else if (webSocketsAvailable && !transportOrOptions.forceHTTP) {
      //All a transport needs to implement is "request"
      this.transport = new WebSocketTransport(endpoint, transportOrOptions);
    } else {
      if (!webSocketsAvailable) {
        console.warn(
          'WebSockets not available, falling back to HTTP transport',
        );
      }

      this.transport = this.http;
    }
  }

  setToken(token: string | Token) {
    this.token = typeof token === 'string' ? fixedToken(token) : token;

    return this.token;
  }

  // NB this is not held anywhere in this instance, but returned
  // to the caller. It only connects when it's subscribe() method
  // is called.
  eventStream(
    filters: string | string[] = [],
    { groupId, offset, scopeId }: Omit<EventStreamRequest, 'filters'> = {},
  ) {
    const filtersArray = Array.isArray(filters) ? filters : [filters];

    if (filtersArray.length === 0) {
      // add a default one that catches everything
      filtersArray.push('(element.ogit/_id = *)');
    }

    return new EventStream(
      { endpoint: this.endpoint, token: this.token },
      { groupId, offset, scopeId, filters: filtersArray },
    );
  }

  /**
   *  Make a clone of this connection, but with new
   */
  cloneWithNewToken(newToken: Token) {
    const cloned = new Client(
      {
        endpoint: this.endpoint,
        token: newToken,
      },
      this.transport,
    );

    return cloned;
  }

  /**
   *  Make a generic request. The transport will handle the details, we handle the retry.
   *  We use `this.transport` which is websocket if available.
   */
  request<T>({ type, headers = {}, body = {} }: GraphRequest, maxRetries = 1) {
    let retries = 0;

    return this.transport
      .request<T>(this.token, { type, headers, body })
      .pipe(
        retryWhen((err$) =>
          err$.pipe(
            concatMap((res) => {
              const closed = res === Errors.connectionClosedBeforeSend;

              // Always handle early close
              if (closed) {
                return of(res);
              }

              if (res.code === 401) {
                this.token.invalidate();
              }

              // Retry retryable error until max retries
              if (
                retries < maxRetries &&
                (res.isRetryable || res.code === 888 || res.code === 401)
              ) {
                retries += 1;

                return of(res);
              }

              return throwError(res);
            }),
            delay(1000),
          ),
        ),
      );
  }

  getToken() {
    return this.token;
  }

  /**
   *  Core GraphIT REST API methods
   *  Prefer websocket transport
   *  The servlets for e.g. /_ki/validate are added using mixins
   */

  /**
   *  Get a single item by ID
   */
  get<T>(id: string, options: { listMeta?: boolean } = {}) {
    const headers = { 'ogit/_id': id } as any;

    if (options.listMeta) {
      headers.listMeta = true;
    }

    return this.request<T>({ type: 'get', headers });
  }

  /**
   *  Get the node for the owner of this token
   */
  me<T>() {
    return this.request<T>({
      type: 'getme',
      body: {},
      headers: { profile: true, 'me-type': 'account' },
    });
  }

  /**
   *  Create a new node
   */
  create<T>(type: string, data = {}, options: { waitForIndex?: boolean } = {}) {
    const headers: Record<string, string> = { 'ogit/_type': type };

    if (options.waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'create', headers, body: data });
  }

  /**
   *  Update a Node
   */
  update<T>(id: string, data = {}, options: { waitForIndex?: boolean } = {}) {
    const headers: Record<string, string> = { 'ogit/_id': id };

    if (options.waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'update', headers, body: data });
  }

  /**
   *  Replace a Node, optionally `upsert`
   */
  replace<T>(
    id: string,
    data: Partial<T> = {},
    options: { waitForIndex?: boolean; createIfNotExists?: string } = {},
  ) {
    const { createIfNotExists = false, waitForIndex = false } = options;
    //createIfNotExists should contain the "ogit/_type" to create if the node doesn't exist,
    //and nothing otherwise
    const headers: Record<string, string> = { 'ogit/_id': id };

    if (createIfNotExists) {
      Object.assign(headers, {
        createIfNotExists: 'true',
        'ogit/_type': createIfNotExists,
      });
    }

    if (waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'replace', headers, body: data });
  }

  /**
   *  Delete a node/edge
   *
   *  We don't handle the 404/409 here. or should we?
   */
  delete<T>(id: string, options: { waitForIndex?: boolean } = {}) {
    const { waitForIndex = false } = options;
    const headers: Record<string, string> = { 'ogit/_id': id };

    if (waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'delete', headers });
  }

  /**
   *  This is a vertices query
   */

  lucene<T>(
    query: string | Lucene.Query = '',
    {
      limit = 50,
      offset = 0,
      order,
      fields = [],
      count = false,
      ...placeholders
    }: LuceneQueryOptions = {},
  ) {
    let _query;
    let _placeholders = placeholders;

    if (typeof query === 'object') {
      const res = CreateLucene(query);

      _query = res.querystring;
      _placeholders = {
        ...placeholders,
        ...res.placeholders,
      };
    } else {
      _query = query;
    }

    const body: {
      query: string;
      limit?: number;
      offset?: number;
      order?: string;
      fields?: string;
      count?: string;
      [index: string]: string | number | undefined;
    } = {
      query: _query,
      limit,
      offset,
      ..._placeholders,
    };

    if (count) {
      body.count = 'true'; // string true
    }

    if (order) {
      body.order = '' + order; // this implicitly does array.join(",") for arrays. (and works with strings...)
    }

    if (fields.length > 0) {
      body.fields = Array.isArray(fields) ? fields.join(',') : String(fields);
    }

    return this.request<T>({
      type: 'query',
      headers: { type: 'vertices' },
      body,
    });
  }

  ids<T>(list: string[]) {
    return this.request<T>({
      type: 'query',
      headers: { type: 'ids' },
      body: { query: list.join(',') }, // yes, it has to be a comma-seperated string
    });
  }

  /**
   *  This is a gremlin query
   */

  gremlin<T>(
    root: string,
    queryOrBuilder: GremlinQuery | GremlinQueryFunction,
  ) {
    const query =
      typeof queryOrBuilder === 'function'
        ? queryOrBuilder(CreateGremlin(''))
        : queryOrBuilder;

    return this.request<T>({
      type: 'query',
      headers: { type: 'gremlin' },
      body: { root, query: query.toString() },
    });
  }

  /**
   *  Connect two Nodes with an edge of `type`
   */
  connect<T>(type: string, inId: string, outId: string) {
    return this.request<T>({
      type: 'connect',
      headers: { 'ogit/_type': type },
      body: { in: inId, out: outId },
    }).pipe(
      catchError((err) => {
        if (Errors.isConflict(err)) {
          return EMPTY;
        }

        throw err;
      }),
      map(() => {}),
    );
  }

  /**
   *  Disconnect two nodes, convenience for delete, generates the edge id for you.
   */
  disconnect<T>(type: string, inId: string, outId: string) {
    return this.delete<T>(`${outId}$$${type}$$${inId}`).pipe(
      catchError((err) => {
        if (Errors.isNotFound(err) || Errors.isConflict(err)) {
          return EMPTY;
        }

        throw err;
      }),
      map(() => {}),
    );
  }

  /**
   *  Write timeseries values (only ogit/Timeseries vertices)
   *
   *  values are { timestamp: millisecond unix, value: string value }
   */
  writets<T>(
    timeseriesId: string,
    values: TimeseriesResponse | TimeseriesResponse[],
  ) {
    const items = Array.isArray(values) ? values : [values];

    return this.request<T>({
      type: 'writets',
      headers: {
        'ogit/_id': timeseriesId,
      },
      body: { items },
    });
  }

  /**
   *  Read timeseries values (only ogit/Timeseries vertices)
   */
  streamts<T>(
    timeseriesId: string,
    options: {
      from?: number | false;
      to?: number | false;
      limit?: number | false;
      includeDeleted?: boolean;
      with?: string[];
    } = {
      from: false,
      to: false,
      limit: 50,
      includeDeleted: false,
      with: [],
    },
  ) {
    const headers = {
      'ogit/_id': timeseriesId,
      ...options,
    };
    const body = {};

    return this.request<T>({
      type: 'streamts',
      headers,
      body,
    });
  }

  /**
   *  Returns previous versions of a vertex
   */
  history<T>(
    id: string,
    {
      listMeta = false,
      includeDeleted = false,
      ...options
    }: HistoryQueryOptions = {},
  ) {
    const headers: Record<string, any> = {
      'ogit/_id': id,
      listMeta,
      includeDeleted,
      ...options,
    };

    return this.request<T>({
      type: 'history',
      headers: headers,
      body: {},
    });
  }
}
