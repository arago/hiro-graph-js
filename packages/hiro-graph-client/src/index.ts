import { Token, cannotGetToken } from './token';
import { Client, ServletFactory } from './client';
import * as Errors from './errors';

export default Client;
export * from './gremlin';
export * from './lucene';
export * from './types';
export * from './jfilter';
export * from './endpoint';
export * as Servlets from './servlets/index';
export { cannotGetToken, Errors, ServletFactory, Token };
