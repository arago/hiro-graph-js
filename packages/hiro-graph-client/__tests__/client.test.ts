// @ts-ignore
import { mockFn } from 'isomorphic-fetch';
import { Subject, BehaviorSubject, of, Observable, throwError } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';

import { Token } from '../src/token';
import * as errors from '../src/errors';
import clientTestHelper from '../__mocks__/client-test-helper';
import Client from '../src';

const getClient = () => {
  const mockRequest = jest.fn();
  const mockInvalidate = jest.fn();
  const token = new Token({
    getToken: () => 'token',
    onInvalidate: mockInvalidate,
  });
  const client = new Client(
    { endpoint: 'https://graphit/', token },
    { forceHTTP: true },
  );

  // @ts-ignore
  client.transport.request = mockRequest;

  return { client, token, mockInvalidate, mockRequest };
};

const enqueueMockResponse = (mock: jest.Mock, ...responses: any[]) => {
  const enqueue = (subscriber: any, res: any) => {
    if (typeof res === 'object') {
      subscriber.error(res);
    } else {
      subscriber.next(res);
    }
  };

  responses.forEach((res) => {
    mock.mockReturnValueOnce(
      new Observable((subscriber) => {
        if (Array.isArray(res)) {
          res.forEach((r) => enqueue(subscriber, r));
        } else {
          enqueue(subscriber, res);
        }

        subscriber.complete();
      }),
    );
  });
};

// test all know good response types. all known error types.
// edge cases like 404, or 409 being OK responses (connect/disconnect)
// bad token causing failure
// etc...
describe('Client Requests', () => {
  const { client, mockRequest } = getClient();

  const tests = clientTestHelper(client);

  beforeEach(() => {
    mockFn.mockReset();
    mockRequest.mockReset();
  });

  tests.forEach(({ name, fn }) => {
    it(`should call ${name} correctly`, async () => {
      //mockFn.mockReturnValueOnce([200, {}]);
      mockRequest.mockReturnValueOnce(of(true));

      // we just have to make sure it resolves period.
      // we will handle expected values for those functions that don't
      // pass the responses right through directly.
      await expect(
        fn()
          .toPromise()
          .then(() => true),
      ).resolves.toEqual(true);
      expect(mockRequest.mock.calls[0][1]).toMatchSnapshot(); //i.e. correct options created
    });
  });
});

describe('Client Response handling', () => {
  const { client, token, mockInvalidate, mockRequest } = getClient();

  beforeEach(async () => {
    mockRequest.mockReset();
    await token.get(); // we want to ensure the token invalidation state is reset each time.
    mockInvalidate.mockClear();
  });

  const retryableErrors = [
    {
      name: 'transaction fail (888)',
      err: errors.transactionFail(),
    },
    {
      name: 'unauthorized (401)',
      err: errors.unauthorized(),
    },
    {
      name: 'generic retryable error',
      err: Object.assign(new Error(), { isRetryable: true }),
    },
  ];

  retryableErrors.forEach(({ name, err }) => {
    it(`should retry once on '${name}'`, async () => {
      enqueueMockResponse(mockRequest, [err, 'ok'], [err, err]);

      await expect(client.me().toPromise()).resolves.toBe('ok');

      await expect(client.me().toPromise()).rejects.toBe(err);
    });
  });

  it('should unconditionally retry for `connection closed before send`', async () => {
    const err = errors.connectionClosedBeforeSend;

    // many connection closed, one other error in the middle, should still resolve "ok"

    enqueueMockResponse(mockRequest, [
      err,
      err,
      err,
      err,
      errors.transactionFail(),
      err,
      err,
      err,
      err,
      'ok',
    ]);

    // @todo How should this be triggered in RxJS?

    await expect(client.me().toPromise()).resolves.toBe('ok');
  });

  it('should handle `conflict` as OK for connect', async () => {
    const conflict = errors.conflict();
    const forbidden = errors.forbidden();

    enqueueMockResponse(mockRequest, conflict, forbidden);

    await expect(
      client.connect('foo', 'bar', 'baz').toPromise(),
    ).resolves.toBeUndefined(); // connect returns nothing on success

    await expect(client.connect('foo', 'bar', 'baz').toPromise()).rejects.toBe(
      forbidden,
    );
  });

  it('should handle `conflict` and `not found` as OK for disconnect', async () => {
    const conflict = errors.conflict();
    const notFound = errors.notFound();
    const forbidden = errors.forbidden();

    enqueueMockResponse(mockRequest, conflict, forbidden, notFound);

    await expect(
      client.disconnect('foo', 'bar', 'baz').toPromise(),
    ).resolves.toBeUndefined(); // disconnect returns nothing on success

    await expect(
      client.disconnect('foo', 'bar', 'baz').toPromise(),
    ).rejects.toBe(forbidden);

    await expect(
      client.disconnect('foo', 'bar', 'baz').toPromise(),
    ).resolves.toBeUndefined(); // disconnect returns nothing on success
  });

  //   const dedupableMethods = [
  //     ['me'],
  //     ['get', 'some-id'],
  //     ['lucene', '*:*'],
  //     ['ids', ['a', 'b']],
  //     ['gremlin', 'root-id', 'outE()'],
  //   ];

  //   dedupableMethods.forEach(([method, ...args]) => {
  //     it(
  //       'should correctly dedup the requests for method: ' + method,
  //       async () => {
  //         let resolve;
  //         const promise = new Promise((r) => (resolve = r));

  //         // note that we only enqueue a single result
  //         // so both client methods MUST only make a single
  //         // request or it will fail Bad Request
  //         client.enqueueMockResponse(promise);

  //         const req1 = client[method](...args);
  //         const req2 = client[method](...args);

  //         resolve({ foo: 'bar' });

  //         const res1 = await req1;
  //         const res2 = await req2;

  //         // these expectations is that bot res1 and res2 returned the same result
  //         // but not the same object.
  //         expect(res1).toEqual(res2);
  //         expect(res1).not.toBe(res2);
  //       },
  //     );
  //   });
  it('should invalidate the token on `unauthorized`', async () => {
    // ok first, then non invalidation error, then invalidation error, then notfound.
    // we need another error after, so the retry can pick it up
    const notFound = errors.notFound();
    const unauthorized = errors.unauthorized();

    enqueueMockResponse(
      mockRequest,
      'OK',
      notFound,
      [unauthorized, notFound],
      'OK',
    );

    await expect(client.me().toPromise()).resolves.toBe('OK');
    expect(mockInvalidate).not.toHaveBeenCalled();

    await expect(client.me().toPromise()).rejects.toBe(notFound);
    expect(mockInvalidate).not.toHaveBeenCalled();

    //this one should invalidate
    await expect(client.me().toPromise()).rejects.toBe(notFound);
    expect(mockInvalidate).toHaveBeenCalledTimes(1);

    await expect(client.me().toPromise()).resolves.toBe('OK');
    expect(mockInvalidate).toHaveBeenCalledTimes(1);
  });
});
