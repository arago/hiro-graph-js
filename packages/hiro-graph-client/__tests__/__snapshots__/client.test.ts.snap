// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Requests should call connect correctly 1`] = `
Object {
  "body": Object {
    "in": "the-in-id",
    "out": "the-out-id",
  },
  "headers": Object {
    "ogit/_type": "some-verb",
  },
  "type": "connect",
}
`;

exports[`Client Requests should call create (waitForIndex) correctly 1`] = `
Object {
  "body": Object {
    "some": "prop",
  },
  "headers": Object {
    "ogit/_type": "some-type",
    "waitForIndex": "true",
  },
  "type": "create",
}
`;

exports[`Client Requests should call create correctly 1`] = `
Object {
  "body": Object {
    "some": "prop",
  },
  "headers": Object {
    "ogit/_type": "some-type",
  },
  "type": "create",
}
`;

exports[`Client Requests should call delete (waitForIndex) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "some-id",
    "waitForIndex": "true",
  },
  "type": "delete",
}
`;

exports[`Client Requests should call delete correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "delete",
}
`;

exports[`Client Requests should call disconnect correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "the-out-id$$some-verb$$the-in-id",
  },
  "type": "delete",
}
`;

exports[`Client Requests should call get correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "get",
}
`;

exports[`Client Requests should call gremlin correctly 1`] = `
Object {
  "body": Object {
    "query": "pow()",
    "root": "root-id",
  },
  "headers": Object {
    "type": "gremlin",
  },
  "type": "query",
}
`;

exports[`Client Requests should call history (with from + to + offset + limit) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "from": 0,
    "includeDeleted": false,
    "limit": 5,
    "listMeta": false,
    "offset": 100,
    "ogit/_id": "some-id",
    "to": 1550597976759,
  },
  "type": "history",
}
`;

exports[`Client Requests should call history (with from + to) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "from": 0,
    "includeDeleted": false,
    "listMeta": false,
    "ogit/_id": "some-id",
    "to": 1550597976759,
  },
  "type": "history",
}
`;

exports[`Client Requests should call history (with from) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "from": 0,
    "includeDeleted": false,
    "listMeta": false,
    "ogit/_id": "some-id",
  },
  "type": "history",
}
`;

exports[`Client Requests should call history (with limit) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "includeDeleted": false,
    "limit": 500,
    "listMeta": false,
    "ogit/_id": "some-id",
  },
  "type": "history",
}
`;

exports[`Client Requests should call history (with offset + limit) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "includeDeleted": false,
    "limit": 5,
    "listMeta": false,
    "offset": 100,
    "ogit/_id": "some-id",
  },
  "type": "history",
}
`;

exports[`Client Requests should call history (with offset) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "includeDeleted": false,
    "listMeta": false,
    "offset": 10,
    "ogit/_id": "some-id",
  },
  "type": "history",
}
`;

exports[`Client Requests should call history (with to) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "includeDeleted": false,
    "listMeta": false,
    "ogit/_id": "some-id",
    "to": 1550597976759,
  },
  "type": "history",
}
`;

exports[`Client Requests should call history (with version) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "includeDeleted": false,
    "limit": 500,
    "listMeta": false,
    "ogit/_id": "some-id",
    "version": 123,
  },
  "type": "history",
}
`;

exports[`Client Requests should call history correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "includeDeleted": false,
    "listMeta": false,
    "ogit/_id": "some-id",
  },
  "type": "history",
}
`;

exports[`Client Requests should call ids correctly 1`] = `
Object {
  "body": Object {
    "query": "a,b,c",
  },
  "headers": Object {
    "type": "ids",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (basic) correctly 1`] = `
Object {
  "body": Object {
    "limit": 50,
    "offset": 0,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (count) correctly 1`] = `
Object {
  "body": Object {
    "count": "true",
    "limit": 50,
    "offset": 0,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (fields) correctly 1`] = `
Object {
  "body": Object {
    "fields": "/field1,/field2",
    "limit": 50,
    "offset": 0,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (limits) correctly 1`] = `
Object {
  "body": Object {
    "limit": -1,
    "offset": 1000,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (order) correctly 1`] = `
Object {
  "body": Object {
    "limit": 50,
    "offset": 0,
    "order": "/field1 asc,/field2 desc",
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (placeholders) correctly 1`] = `
Object {
  "body": Object {
    "bar": "123",
    "limit": 50,
    "offset": 0,
    "query": "/foo:$bar",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call me correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "me-type": "account",
    "profile": true,
  },
  "type": "getme",
}
`;

exports[`Client Requests should call replace (createIfNotExists) correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "createIfNotExists": "true",
    "ogit/_id": "some-id",
    "ogit/_type": "some-type",
  },
  "type": "replace",
}
`;

exports[`Client Requests should call replace (waitForIndex) correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
    "waitForIndex": "true",
  },
  "type": "replace",
}
`;

exports[`Client Requests should call replace correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "replace",
}
`;

exports[`Client Requests should call streamts (from only) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "from": 1399999999999,
    "ogit/_id": "ts-id",
  },
  "type": "streamts",
}
`;

exports[`Client Requests should call streamts (to + from) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "from": 1399999999999,
    "ogit/_id": "ts-id",
    "to": 1400000000001,
  },
  "type": "streamts",
}
`;

exports[`Client Requests should call streamts (to only) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "ts-id",
    "to": 1400000000001,
  },
  "type": "streamts",
}
`;

exports[`Client Requests should call streamts correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "from": false,
    "includeDeleted": false,
    "limit": 50,
    "ogit/_id": "ts-id",
    "to": false,
    "with": Array [],
  },
  "type": "streamts",
}
`;

exports[`Client Requests should call update (waitForIndex) correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
    "waitForIndex": "true",
  },
  "type": "update",
}
`;

exports[`Client Requests should call update correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "update",
}
`;

exports[`Client Requests should call writets correctly 1`] = `
Object {
  "body": Object {
    "items": Array [
      Object {
        "timestamp": 1400000000000,
        "value": "value1",
      },
    ],
  },
  "headers": Object {
    "ogit/_id": "ts-id",
  },
  "type": "writets",
}
`;
